druid/examples/calc.rs

```diff
@@ -112,95 +115,239 @@ impl CalcState {
+            KeyCode::Numpad1 | KeyCode::Key1 => Some(Digit(1)),
+            KeyCode::Numpad2 | KeyCode::Key2 => Some(Digit(2)),
+            KeyCode::Numpad3 | KeyCode::Key3 => Some(Digit(3)),
+            KeyCode::Numpad4 | KeyCode::Key4 => Some(Digit(4)),
+            KeyCode::Numpad5 | KeyCode::Key5 => Some(Digit(5)),
+            KeyCode::Numpad6 | KeyCode::Key6 => Some(Digit(6)),
+            KeyCode::Numpad7 | KeyCode::Key7 => Some(Digit(7)),
+            KeyCode::Numpad8 | KeyCode::Key8 => Some(Digit(8)),
+            KeyCode::Numpad9 | KeyCode::Key9 => Some(Digit(9)),
+            KeyCode::NumpadSubtract | KeyCode::Minus => Some(Op('−')),
+            KeyCode::NumpadAdd => Some(Op('+')),
+            KeyCode::NumpadMultiply => Some(Op('×')),
+            KeyCode::NumpadEnter | KeyCode::NumpadEquals | KeyCode::Equals | KeyCode::Return => {
+                Some(Op('='))
+            }
+            KeyCode::Slash | KeyCode::NumpadDivide => Some(Op('÷')),
+            KeyCode::KeyC => Some(Op('c')),
+            KeyCode::Backspace => Some(Op('⌫')),
+            KeyCode::NumpadDecimal | KeyCode::Period => Some(Op('.')),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+impl Widget<CalcState> for KeyMap {
+    fn paint(
+        &mut self,
+        paint_ctx: &mut PaintCtx,
+        base_state: &BaseState,
+        data: &CalcState,
+        env: &Env,
+    ) {
+        self.flex.paint(paint_ctx, base_state, data, env)
+    }
+
+    fn layout(
+        &mut self,
+        layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        data: &CalcState,
+        env: &Env,
+    ) -> Size {
+        self.flex.layout(layout_ctx, bc, data, env)
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut CalcState, env: &Env) {
+        use crate::ButtonId::*;
+        match event {
+            Event::KeyDown(key_event) => {
+                if key_event.is_repeat {
+                    return;
+                };
+
+                let kc = key_event_to_button_id(key_event);
+                // Current we can just perform the action on the CalcState,
+                // However we don't get the pretty button animation this way.
+                match kc {
+                    None => self.flex.event(ctx, event, data, env),
+                    Some(Op(c)) => data.op(c),
+                    Some(Digit(d)) => data.digit(d),
+                }
+                match kc {
+                    None => (),
+                    Some(button_id) => {
+                        let ButtonInfo {
+                            row_id: _row_id,
+                            col_id: _col_id,
+                        } = self.key_map.get(&button_id).unwrap();
+                        // In theory if there was a mechanism to get random access into
+                        // self.flex.children we could remove the above match,
+                        // and do something fancy with the buttons here instead.
```

Comment:
there's a maybe-hacky-maybe-just-how-we-do-things way to do this (although it would require some new logic in button): basically you would define a `Command` for each button, and when the hotkey was triggered you would send the appropriate `Command`; each button would be listening for its own command, and when it was received it would trigger the button's action and do the animation, etc.